### DISCUSSION TITLE
next/image with remote URL return 504 error

### DISCUSSION THREAD

**Discussion by dmitrydrynov** [vercel/next.js#34433] [Help]
Hi everyone.
I need help )

In my project I have:
backend docker container: https://api.somesite.com (fastify) 
frontend docker container: https://somesite.com (next.js)

When I use internal images with next/image all ok (I see the image)
`https://somesite.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Flogo.svg&w=640&q=75`

This image opens well
`https://api.somesite.com/storage/avatars/avatar-y7aADf1EqzGA98eg.png`

When I use remote images with next/image I'm getting 504 error
`https://somesite.com/_next/image?url=https%3A%2F%2Fapi.somesite.com%2Fstorage%2Favatars%2Favatar-y7aADf1EqzGA98eg.png&w=256&q=75`

```
FetchError: request to https://api.somesite.com/storage/avatars/avatar-j6e9rfY27yLmpDaW.png failed, reason: connect ETIMEDOUT 87.252.254.0:443
    at ClientRequest.<anonymous> (/app/node_modules/next/dist/compiled/node-fetch/index.js:1:64142)
    at ClientRequest.emit (node:events:520:28)
    at TLSSocket.socketErrorListener (node:_http_client:442:9)
    at TLSSocket.emit (node:events:520:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  type: 'system',
  errno: 'ETIMEDOUT',
  code: 'ETIMEDOUT'
```

The backend hostname I added to
```
module.exports = {
  images: {
    domains: ['somesite.com'],
  },
}
```

**Comment by 2stash** [comment]
@dmitrydrynov Hello. Have you tried adding different domains to the domains export? Specifically your error says you cannot reach the image at "https://api.skillkit.me" which is not 'somesite.com'.

**Reply by dmitrydrynov** [reply]
@2stash [2stash](https://github.com/2stash)
api.somesite.com equal api.skillkit.me
I wrote somesite.com as example. 

p.s. updated the first post

**Reply by 2stash** [reply]
@dmitrydrynov Ok. I just did a test and you will get an error if the external domain is not properly set in your next.config.js file. Since you are getting a 504 error, everything with your config should be good, so I would recommend doing two things. 

First I would put that image in a normal HTML image tag. Using the next.js Image component is for sizing and optimization which is not required. This will help point to the correct problem because it could be the link itself or the server you are trying to get it from. Second I would try using the Image tag with a random external image. I was just using https://picsum.photos/ to experiment with the Image component.

If you've already done these things or they don't help. Can you add some of your code, specifically it would help to see your Image component and how it is rendered.

**Reply by dmitrydrynov** [reply]
> @dmitrydrynov Ok. I just did a test and you will get an error if the external domain is not properly set in your next.config.js file. Since you are getting a 504 error, everything with your config should be good, so I would recommend doing two things.
> 
> First I would put that image in a normal HTML image tag. Using the next.js Image component is for sizing and optimization which is not required. This will help point to the correct problem because it could be the link itself or the server you are trying to get it from. Second I would try using the Image tag with a random external image. I was just using https://picsum.photos/ to experiment with the Image component.
> 
> If you've already done these things or they don't help. Can you add some of your code, specifically it would help to see your Image component and how it is rendered.

I tried with cdn.discordapp.com domain, all ok
backend and frontend docker containers are currently on the same server

**Reply by 2stash** [reply]
@dmitrydrynov What next.js version are you using?

**Reply by dmitrydrynov** [reply]
Next.js 12.0.10

I think my problem is in docker settings. 
I need to redirect the external host in the frontend container from api.somesite.com:433 to backend_container_name:8000

**Comment by ribeaud** [comment]
We are experiencing a similar problem on our side (a more detailed bug report could be found [here](https://gitlab.com/biomedit/portal/-/issues/537)).

The bug is really hard to reproduce. However, after a while (and this is reproducible by just clicking around in the application), it randomly starts to trigger `504`(_bad gateway_) errors. First the images, then the problem propagates to the whole application. My guess is that the problem is related to `next/image` (and because _v12.0.10_ brings a lot of issues related to `next/image`).

While trying and narrowing down the problem, I've found out that downgrading to _v12.0.9_ seems to solve the problem. I did not create a bug report due to the nature of the problem (as I said, really hard to reproduce).

**Reply by 2stash** [reply]
@ribeaud If you don't see the issue again after changing the next.js version that is a strong case for a bug. I checked through the open issues and I didn't see one that matched this, or was this severe. Have you done any other troubleshooting? Please share if you did. It would be interesting to see if you stopped using the Image components gets rid of the problem. The other question I have is aren't these images cached by react? So clicking around the app shouldn't reload the images? Or is it more that you have LOTS of images to load and after you've loaded so many the images stop working and the app starts to crash?  

Please follow up if your issue doesn't return on v12.0.9 as that would be the answer.

**Reply by ribeaud** [reply]
Good morning. No, we do not have a lot of images. We will investigate further, but beginning of next week. To make a bug report, I have to try out **canary** version first, correct? And you are suggesting that we should check:
- Whether the images are cached by **React**
- Whether NOT using the image component solves the problem

**Reply by ribeaud** [reply]
@2stash How about [that](https://github.com/vercel/next.js/issues/33860) one? We will try **canary** out next week and will let you know...

**Reply by 2stash** [reply]
Bonjour! That issue looks pretty close!

Yes you could test things out on a canary branch. But I meant to try the easiest things first. It might help out to do or think about the following and add it to that open issue.

- Change <Image ... /> to standard HTML <img ... />. Since you don't have hundreds of images this should be quick to do. If you use standard HTML and still have a problem, then it's probably your code/config that is wrong. (doing this might break your styling, so the page might look ugly with huge images, but as long as they load should be ok. Or you could inline style the img tags)
- React is a SPA, single page application. Meaning if you click on a page, everything should load, and things that are not changing should not load again. So I don't think you would want your images to constantly reload as a user clicks around. They should only reload if the page refreshes or if the images change. Next.js gives you more control over how and when your data refreshes. So I would just make sure that in your code, the images only load the first time a user goes to the page. (unless they are dynamic and should change). You wrote "However, after a while (and this is reproducible by just clicking around in the application)". So I'm not sure if you mean "I click around ten times and it breaks", or "I click around really fast 100 times and it breaks". 

So if user clicks on your Homepage, and images load, then they click on another page and that page and images load. Then they click on the Homepage again, the images should not have to reload, they should already be loaded. But if you have it this way so the Homepage images don't reload, but they break anyways... Just trying to figure out what exactly is happening when the app stops working.
-does using v12.0.9 for a week solve the problem.

**Reply by Jaynam07** [reply]
Similar issue - on local its working properly, all images are being loaded lazily and properly and even while running production build on local.

But while deploying it on - Docker / Kubernetes /Nginx - few images are not getting loaded and after some time it shows 504.
can some one help here ?

**Reply by ribeaud** [reply]
The issue disappeared in the recent **Next.js** releases. There was a race condition in images cache, for more details, kindly refer to https://github.com/vercel/next.js/issues/33860.

**Reply by 44dhi** [reply]
I am still facing the same issue, on local everything runs fine, but when I deploy it on a digital ocean container, most of the images don't load and I face a 504 error, I'm on NEXT js 12.1.5

**Reply by Chanaka-Sandaruwan** [reply]
I also have same issue, I am on NEXT js 12.1.5

**Reply by arhamymr** [reply]
how to fix this issue  without downgrade to 12.0.9 ? i have it to in 12.3.1

**Comment by mohamed-badraldin** [comment]
<img src={user?.image} /> ❌
<img src={user.image} /> ✔

**Reply by absolutegravitas** [reply]
still an issue in 2024 - how this package keeps being used by people is beyond me...
